Add deathrattle implant for nuclear operatives

coiax opened the pull request:
:cl: coiax
add: Nuke ops can now purchase a box of "deathrattle implants". When an
implanted person dies, all the other users of the implant will get a
message, saying who died and where they died.
/:cl:

Deathrattle implant, comes in box of 8, implant yourself and others, and
you'll get a little notification when they die.

Intended for nuke ops, price around the same as the codebook, since it's
a utility communication item, that actually "plans to fail", so it can't
be too expensive, because it doesn't help them win, just lets them know
how much they are losing by.

---

- [x] Actually add to nuke op uplink
- [x] Figure out why implant doesn't fire if it's implanted after an explosive implant.

coiax committed:
Add deathrattle implant for nuclear operatives

:cl: coiax
add: Nuke ops can now purchase a box of "deathrattle implants". When an
implanted person dies, all the other users of the implant will get a
message, saying who died and where they died.
/:cl:

Deathrattle implant, comes in box of 8, implant yourself and others, and
you'll get a little notification when they die.

Intended for nuke ops, price around the same as the codebook, since it's
a utility communication item, that actually "plans to fail", so it can't
be too expensive, because it doesn't help them win, just lets them know
how much they are losing by.

Dorsisdwarf commented:
Have you considered that the implant isn't firing because when someone dies the explosive implant goes off and destroys the rattle implant before it can trigger because its first in the implant list ?

coiax committed:
Change deathrattle implant to use PREDEATH signal

- To ensure that the deathrattle fires, we fire it before the other
  LIVING_DEATH signals, the most famous of which includes the explosive
  implant, which will destroy the body (and the other implants) when it
  runs.

- Added the deathrattle implant collection to the syndicate uplink at
  4TC. The equivalent item, is the Codespeak Manual, which costs 3TC,
  and effectively makes your communications secure when one of you dies
  and leaks a headset. Both utility items, that don't have any benefit
  unless you start to lose really.

coiax commented:
> 
> 
> Have you considered that the implant isn't firing because when someone dies the explosive implant goes off and destroys the rattle implant before it can trigger because its first in the implant list ?

Yeah, that is the most likely explanation.

san7890 commented:
Does this work on Cayenne?

coiax commented:
> 
> 
> Does this work on Cayenne?

Yes. It working on Cayenne is an important part of the featureset. 🐟

Cyberboss commented on code/modules/mob/living/death.dm:
```diff
@@ -69,6 +69,8 @@
  * * gibbed - Was the mob gibbed?
 */
 /mob/living/proc/death(gibbed)
+	SEND_SIGNAL(src, COMSIG_LIVING_PREDEATH, gibbed)
```
@ninjanomnom I wanna know if you'd rather see this handled some other way. See the commit message of https://github.com/tgstation/tgstation/pull/55903/commits/ba3099b1f8b6e4bd15b2e5629a3e1b941213ab6d

Rohesie replied:
Do you want to use the `gibbed` value? Else `set_stat` sends a signal already.

Cyberboss replied:
If that can be done I think it would be preferable. This seems heavy-handed.

coiax committed:
Merge branch 'master' into nukeop-has-died-in-infiltrator

coiax committed:
Uses COMSIG_MOB_STATCHANGE signal

Instead of making a new signal that fires before the rest of the death
code, we can use COMSIG_MOB_STATCHANGE, which is sent first, before the
rest of the death code.

Rohesie commented on code/game/objects/items/storage/uplink_kits.dm:
```diff
@@ -565,3 +565,20 @@
+
+	var/implants = list()
+	for(var/j in 1 to 8)
+		var/obj/item/implantcase/deathrattle/case = new /obj/item/implantcase/deathrattle(src)
```
```suggestion
		var/obj/item/implantcase/deathrattle/case = new (src)
```
Already typed by the variable.

coiax committed:
Update code/game/objects/items/storage/uplink_kits.dm

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/game/objects/items/implants/implant_deathrattle.dm:
```diff
@@ -0,0 +1,69 @@
+		RegisterSignal(target, COMSIG_MOB_STATCHANGE, .proc/on_statchange)
+
+		if(!group)
+			var/msg = "<i>You hear a strange, robotic voice in your head...</i> \"<span class='robot'>Warning: No other linked implants detected.</span>\""
+			to_chat(target, msg)
```
```suggestion
			to_chat(target, "<i>You hear a strange, robotic voice in your head...</i> \"<span class='robot'>Warning: No other linked implants detected.</span>\"")
```
Feel free to ignore this if you think it's less clear, but `msg` seems to suggest me that it will be modified or used more than once, and since it is not it kind of defeats the purpose.

Rohesie commented on code/game/objects/items/implants/implant_deathrattle.dm:
```diff
@@ -0,0 +1,69 @@
+
+/obj/item/implant/deathrattle/implant(mob/living/target, mob/user, silent = FALSE, force = FALSE)
+	. = ..()
+	if(.)
+		RegisterSignal(target, COMSIG_MOB_STATCHANGE, .proc/on_statchange)
```
Do an early return instead.

Rohesie commented on code/game/objects/items/implants/implant_deathrattle.dm:
```diff
@@ -0,0 +1,69 @@
+
+/obj/item/implant/deathrattle/Destroy()
+	group = null
+	. = ..()
```
```suggestion
	return ..()
```

coiax committed:
Merge branch 'master' into nukeop-has-died-in-infiltrator

coiax committed:
Apply suggestions from code review

coiax committed:
Merge branch 'master' into nukeop-has-died-in-infiltrator

Rohesie commented:
Will you move the deathrattle to a component/element like we talked about on discord, to avoid circular references?

coiax committed:
Rework signal ownership to group

Instead of circular references between the group and the implant, the
implant now does nothing on its own, it's the group that registers the
signals that make the whole system works.
